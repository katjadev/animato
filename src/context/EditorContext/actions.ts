import { Dispatch } from 'react'
import { Action } from '@animato/types'

export const actionTypes = {
  START_PLAYING: 'START_PLAYING',
  STOP_PLAYING: 'STOP_PLAYING',
  TOGGLE_REPEAT_MODE: 'TOGGLE_REPEAT_MODE',
  HOVER_ELEMENT: 'HOVER_ELEMENT',
  TOGGLE_ELEMENT: 'TOGGLE_ELEMENT',
  SET_CURRENT_TIME: 'SET_CURRENT_TIME',
  SET_TIMELINE_WIDTH: 'SET_TIMELINE_WIDTH',
  SET_SCROLL_POSITION: 'SET_SCROLL_POSITION',
  SET_ZOOM: 'SET_ZOOM',
  COLLAPSE_ELEMENT: 'COLLAPSE_ELEMENT',
  EXPAND_ELEMENT: 'EXPAND_ELEMENT',
  COLLAPSE_ANIMATION: 'COLLAPSE_ANIMATION',
  EXPAND_ANIMATION: 'EXPAND_ANIMATION',
}

export const editorActions = (dispatch: Dispatch<Action>) => (
  {
    startPlaying: () => dispatch({ type: actionTypes.START_PLAYING }),
    stopPlaying: () => dispatch({ type: actionTypes.STOP_PLAYING }),
    toggleRepeatMode: () => dispatch({ type: actionTypes.TOGGLE_REPEAT_MODE }),
    hoverElement: (payload: { id: string | null }) => dispatch({ type: actionTypes.HOVER_ELEMENT, payload }),
    toggleElement: (payload: { id: string, multiple?: boolean }) => dispatch({ type: actionTypes.TOGGLE_ELEMENT, payload }),
    setCurrentTime: (payload: { value: number }) => dispatch({ type: actionTypes.SET_CURRENT_TIME, payload }),
    setTimelineWidth: (payload: { width: number }) => dispatch({ type: actionTypes.SET_TIMELINE_WIDTH, payload }),
    setScrollPosition: (payload: { value: number }) => dispatch({ type: actionTypes.SET_SCROLL_POSITION, payload }),
    setZoom: (payload: { value: number }) => dispatch({ type: actionTypes.SET_ZOOM, payload }),
    collapseElement: (payload: { id: string }) => dispatch({ type: actionTypes.COLLAPSE_ELEMENT, payload }),
    expandElement: (payload: { id: string }) => dispatch({ type: actionTypes.EXPAND_ELEMENT, payload }),
    collapseAnimation: (payload: { id: string }) => dispatch({ type: actionTypes.COLLAPSE_ANIMATION, payload }),
    expandAnimation: (payload: { id: string }) => dispatch({ type: actionTypes.EXPAND_ANIMATION, payload }),
  }
)
