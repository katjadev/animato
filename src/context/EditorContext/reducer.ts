import { Action, ScrollPosition, TimelineMark } from '@animato/types'
import { MAX_ZOOM, REM_TO_PX_COEFFICIENT, TIMELINE_PADDING } from '@animato/constants'
import generateTimelineMarks from '@animato/utils/generateTimelineMarks'
import { actionTypes } from './actions'

export type EditorState = {
  isPlaying: boolean;
  isRepeatMode: boolean;
  hoveredElementId: string | null;
  selectedElementIds: string[];
  currentTime: number;
  timelineWidth: number;
  scrollPosition: ScrollPosition;
  zoom: number;
  timelineMarks: TimelineMark[];
  collapsedElements: string[];
  collapsedAnimations: string[];
}

const timelinePaddingPx = TIMELINE_PADDING * REM_TO_PX_COEFFICIENT
const timelineMarks = generateTimelineMarks(MAX_ZOOM)
const markSize = timelineMarks[1].position - timelineMarks[0].position
const timelineWidth = timelineMarks.length * markSize + 2 * timelinePaddingPx

export const initialEditorState: EditorState = {
  isPlaying: false,
  isRepeatMode: false,
  hoveredElementId: null,
  selectedElementIds: [],
  currentTime: 0,
  timelineWidth,
  scrollPosition: { top: 0, left: 0 },
  zoom: MAX_ZOOM,
  timelineMarks,
  collapsedElements: [],
  collapsedAnimations: [],
}

export const editorReducer = (state: EditorState, action: Action) => {
  switch (action.type) {
    case actionTypes.START_PLAYING:
      return {
        ...state,
        isPlaying: true,
      }
    case actionTypes.STOP_PLAYING:
      return {
        ...state,
        isPlaying: false,
      }
    case actionTypes.TOGGLE_REPEAT_MODE:
      return {
        ...state,
        isRepeatMode: !state.isRepeatMode,
      }
    case actionTypes.HOVER_ELEMENT:
      return {
        ...state,
        hoveredElementId: action.payload!.id,
      }
    case actionTypes.SELECT_ELEMENT:
      return {
        ...state,
        selectedElementIds: [...state.selectedElementIds, action.payload!.id],
      }
    case actionTypes.DESELECT_ELEMENT:
      return {
        ...state,
        selectedElementIds: state.selectedElementIds.filter(id => id !== action.payload!.id),
      }
    case actionTypes.SET_CURRENT_TIME:
      return {
        ...state,
        currentTime: action.payload!.value,
      }
    case actionTypes.SET_TIMELINE_WIDTH:
      return {
        ...state,
        timelineWidth: action.payload!.width,
      }
    case actionTypes.SET_SCROLL_POSITION:
      return {
        ...state,
        scrollPosition: action.payload!.value,
      }
    case actionTypes.SET_ZOOM:
      const zoom = action.payload!.value
      const timelineMarks = generateTimelineMarks(zoom)
      const markSize = timelineMarks[1].position - timelineMarks[0].position
      const timelineWidth = timelineMarks.length * markSize + 2 * timelinePaddingPx
      return {
        ...state,
        zoom,
        timelineMarks,
        timelineWidth,
      }
    case actionTypes.COLLAPSE_ELEMENT:
      return {
        ...state,
        collapsedElements: [...state.collapsedElements, action.payload!.id],
      }
    case actionTypes.EXPAND_ELEMENT:
      return {
        ...state,
        collapsedElements: state.collapsedElements.filter(id => id !== action.payload!.id),
      }
    case actionTypes.COLLAPSE_ANIMATION:
      return {
        ...state,
        collapsedAnimations: [...state.collapsedAnimations, action.payload!.id],
      }
    case actionTypes.EXPAND_ANIMATION:
      return {
        ...state,
        collapsedAnimations: state.collapsedAnimations.filter(id => id !== action.payload!.id),
      }
    default:
      return state
  }
}
