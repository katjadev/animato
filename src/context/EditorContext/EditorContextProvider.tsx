import { ReactNode, createContext, useContext, useReducer } from 'react'
import { EditorState, editorReducer, initialEditorState } from './reducer'
import { editorActions } from './actions'



export const EditorContext = createContext<{
  state: EditorState;
  actions: any;
}>({
  state: initialEditorState,
  actions: {},
})

interface EditorContextProviderProps {
  children: ReactNode;
}

export default function EditorContextProvider({children}: EditorContextProviderProps) {
  const [state, dispatch] = useReducer(editorReducer, initialEditorState)
  const actions = editorActions(dispatch)

  return (
    <EditorContext.Provider value={{ state, actions }}>
      {children}
    </EditorContext.Provider>
  )
}

export const useEditorState = () => useContext(EditorContext)