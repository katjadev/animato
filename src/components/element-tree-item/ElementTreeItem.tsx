import { FC, useState } from 'react'
import { ElementTreeNode } from '@animato/types'
import { useEditorState } from '@animato/context/EditorContext/EditorContextProvider'
import { useProjectState } from '@animato/context/ProjectContext/ProjectContextProvider'
import Icon from '@animato/components/icon/Icon'
import IconButton from '../icon-button/IconButton'
import Menu from '../menu/Menu'
import MenuItem from '../menu-item/MenuItem'
import styles from './ElementTreeItem.module.css'
import EditableText from '../editable-text/EditableText'

const icons: { [key: string]: string } = {
  rect: 'square',
  circle: 'circle',
  g: 'intersect',
  path: '3d-bridge',
}

export type ElementTreeItemTranslations = {
  expand: string;
  collapse: string;
  elementAriaLabel: string;
}

interface ElementTreeItemProps {
  element: ElementTreeNode;
  level: number;
  selected: boolean;
  translations: ElementTreeItemTranslations;
}

const ElementTreeItem: FC<ElementTreeItemProps> = ({
  element, 
  level, 
  selected, 
  translations,
}) => {
  const { actions: projectActions } = useProjectState()
  const { state, actions: editorActions } = useEditorState()
  const { collapsedElements } = state

  const icon = icons[element.tagName] || 'flare'
  const isCollapsed = collapsedElements.includes(element.id)

  const select = () => editorActions.selectElement({ id: element.id })
  const deselect = () => editorActions.selectElement({ id: null })
  const collapse = () => editorActions.collapseElement({ id: element.id })
  const expand = () => editorActions.expandElement({ id: element.id })

  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null)
  const open = Boolean(anchorEl)

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  const handleRename = (value: string) => {
    projectActions.renameElement({ id: element.id, value })
  }

  return (
    <div 
      className={`${styles.title} ${selected ? styles.selected :''}`} 
      tabIndex={0}
      onMouseEnter={select}
      onMouseLeave={deselect}
      onFocus={select}
      onBlur={deselect}
    >
      <div className={`${styles.titleIn} ${styles[`level${level}`]}`}>
        {element.children.length > 0 && (
          <>
            {isCollapsed ? (
              <button
                className={styles.collapseButton}
                aria-label={translations.expand}
                onClick={expand}
              >
                <Icon icon='nav-arrow-right' />
              </button>
            ) : (
              <button
                className={styles.collapseButton}
                aria-label={translations.collapse}
                onClick={collapse}
              >
                <Icon icon='nav-arrow-down' />
              </button>
            )}
          </>
        )}
        <Icon className={styles.shapeIcon} icon={icon} />
        <div className={styles.titleText}>
          <EditableText
            value={element.title}
            aria-label={'Rename element'}
            onChange={handleRename}
          />
        </div>
        <IconButton 
          aria-label={translations.elementAriaLabel} 
          size='small'
          onClick={handleClick}
        >
          <Icon icon='more-vert' />
        </IconButton>
        <Menu
          id='element-menu'
          anchorEl={anchorEl}
          open={open}
          onClose={handleClose}
        >
          <MenuItem onClick={() => { console.log('!!! animate') }}>
            <Icon icon='bounce-right' />
            Animate
          </MenuItem>
        </Menu>
      </div>
    </div>
  )
}

export default ElementTreeItem