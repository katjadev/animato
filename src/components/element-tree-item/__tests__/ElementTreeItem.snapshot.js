import { render } from '@testing-library/react'
import { useEditorState } from '@animato/context/EditorContext/EditorContextProvider'
import ElementTreeItem from '../ElementTreeItem'

jest.mock('@animato/context/EditorContext/EditorContextProvider', () => ({
  useEditorState: jest.fn().mockReturnValue({
    state: {
      collapsedElements: [],
    },
    actions: {},
  })
}))

const translations = {
  expand: 'Expand',
  collapse: 'Collapse',
}

const TEST_ID = 'test-id'
const testElement = {
  id: 'test-id',
  title: 'test-element',
  element: document.createElement('rect'),
  children: [],
}

describe('ElementTreeItem', () => {
  it('renders correctly for a leaf element', () => {
    const { container } = render(
      <ElementTreeItem
        level={0}
        element={testElement}
        translations={translations}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders correctly for a parent element', () => {
    const { container } = render(
      <ElementTreeItem
        level={0}
        element={{
          ...testElement,
          children: [{
            id: 'test-child-id',
            title: 'test-child-element',
            element: document.createElement('path'),
            children: [],
          }],
        }}
        translations={translations}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders correctly for a collapsed parent element', () => {
    const TEST_ID = 'test-id'
    useEditorState.mockReturnValue({
      state: {
        collapsedElements: [TEST_ID],
      },
      actions: {},
    })

    const { container } = render(
      <ElementTreeItem
        level={0}
        element={{
          ...testElement,
          children: [{
            id: 'test-child-id',
            title: 'test-child-element',
            element: document.createElement('path'),
            children: [],
          }],
        }}
        translations={translations}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
