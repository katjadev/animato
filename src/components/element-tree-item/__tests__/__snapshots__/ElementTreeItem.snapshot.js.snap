// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementTreeItem renders correctly for a collapsed parent element 1`] = `
<div>
  <div
    class="title"
    tabindex="0"
  >
    <div
      class="titleIn level0"
    >
      <button
        aria-label="Expand"
        class="collapseButton"
      >
        <i
          class="iconoir-nav-arrow-right "
        />
      </button>
      <div>
        test-element
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementTreeItem renders correctly for a leaf element 1`] = `
<div>
  <div
    class="title"
    tabindex="0"
  >
    <div
      class="titleIn level0"
    >
      <div>
        test-element
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementTreeItem renders correctly for a parent element 1`] = `
<div>
  <div
    class="title"
    tabindex="0"
  >
    <div
      class="titleIn level0"
    >
      <button
        aria-label="Collapse"
        class="collapseButton"
      >
        <i
          class="iconoir-nav-arrow-down "
        />
      </button>
      <div>
        test-element
      </div>
    </div>
  </div>
</div>
`;
